core/templates/dev/head/pages/exploration_player/ImagePreloaderService.js

```diff
@@ -0,0 +1,198 @@
+    // directive).The object contains the resolve method of the promise
+    // attached with getInImageUrl method.
+    var _imageLoadedCallback = {};
+
+    var _init = function(exploration) {
+      _exploration = exploration;
+      _states = exploration.states;
+    };
+
+    /**
+     * Gets the Url for the image file.
+     * @param {string} filename - Filename of the image whose Url is to be
+     *                            created.
+     * @param {function} onLoadCallback - Function that is called when the
+     *                                    Url of the loaded image is obtained.
+     */
+    var _getUrlUsingFileInCache = function(filename, onLoadCallback) {
+      AssetsBackendApiService.loadImage(
+        ExplorationContextService.getExplorationId(), filename)
+        .then(function(loadedImageFile) {
+          var objectUrl = URL.createObjectURL(loadedImageFile.data);
+          onLoadCallback(objectUrl);
+        });
+    };
+
+    /**
+     * Gets image files in Bfs order from the state.
+     * @param {string} sourceStateName - The name of the state starting from
+     *                                   which the filenames should be
+     *                                   obtained.
+     */
+    var _getImageFilenamesInBfsOrder = function(sourceStateName) {
+      var stateNamesInBfsOrder = (
+        ComputeGraphService.computeBfsTraversalOfStates(
+          _exploration.getInitialState().name,
+          _exploration.getStates(),
+          sourceStateName));
+      var imageFilenames = [];
+
+      stateNamesInBfsOrder.forEach(function(stateName) {
+        var state = _states.getState(stateName);
+        ExtractImageFilenamesFromStateService.getImageFilenamesInState(state).
+          forEach(function(filename) {
+            imageFilenames.push(filename);
+          });
+      });
+      return imageFilenames;
+    };
+
+    /**
+     * Handles the loading of the image file.
+     * @param {string} imageFilename - The filename of the image to be loaded.
+     */
+    var _loadImage = function(imageFilename) {
+      AssetsBackendApiService.loadImage(
+        ExplorationContextService.getExplorationId(), imageFilename
+      ).then(function(loadedImage) {
+        for (var i = 0;
+          i < _filenamesOfImageCurrentlyDownloading.length; i++) {
+          if (_filenamesOfImageCurrentlyDownloading[i] ===
+              loadedImage.filename) {
+            _filenamesOfImageCurrentlyDownloading.splice(i, 1);
+            break;
+          }
+        }
+        if (_filenamesOfImageToBeDownloaded.length > 0) {
+          var nextImageFilename = _filenamesOfImageToBeDownloaded.shift();
+          _filenamesOfImageCurrentlyDownloading.push(nextImageFilename);
+          _loadImage(nextImageFilename);
+        }
+        if (_imageLoadedCallback[loadedImage.filename]) {
+          var onLoadImageResolve = (
+            (_imageLoadedCallback[loadedImage.filename]).resolveMethod);
```

Comment:
What I mean is that `_imageLoadedCallback[loadedImage.filename].resolveMethod` should work the same.

Reply:
No, it would give an error that the _imageLoadedCallback[loadedImage.filename] is undefined. so we can't use the whole _imageLoadedCallback[loadedImage.filename].resolveMethod